python/ray/data/_internal/execution/bulk_executor.py

```diff
@@ -0,0 +1,109 @@
+    def __init__(self, options: ExecutionOptions):
+        super().__init__(options)
+        self._stats: Optional[DatasetStats] = DatasetStats(stages={}, parent=None)
+        self._executed = False
+
+    def execute(
+        self, dag: PhysicalOperator, initial_stats: Optional[DatasetStats] = None
+    ) -> Iterator[RefBundle]:
+        """Synchronously executes the DAG via bottom-up recursive traversal."""
+
+        assert not self._executed, "Can only call execute once."
+        self._executed = True
+        if not isinstance(dag, InputDataBuffer):
+            logger.info("Executing DAG %s", dag)
+
+        if initial_stats:
+            self._stats = initial_stats
+
+        saved_outputs: Dict[PhysicalOperator, List[RefBundle]] = {}
+
+        def execute_recursive(op: PhysicalOperator) -> List[RefBundle]:
+            # Avoid duplicate executions.
+            if op in saved_outputs:
+                return saved_outputs[op]
+
+            # Compute dependencies.
+            inputs = [execute_recursive(dep) for dep in op.input_dependencies]
+
+            # Fully execute this operator.
+            logger.debug("Executing op %s", op.name)
+            builder = self._stats.child_builder(op.name)
+            try:
+                for i, ref_bundles in enumerate(inputs):
+                    for r in ref_bundles:
+                        op.add_input(r, input_index=i)
+                    op.inputs_done(i)
+                output = _naive_run_until_complete(op)
+            finally:
+                op.shutdown()
+
+            # Cache and return output.
+            saved_outputs[op] = output
+            op_stats = op.get_stats()
+            op_metrics = op.get_metrics()
+            if op_stats:
+                self._stats = builder.build_multistage(op_stats)
+                self._stats.extra_metrics = op_metrics
+            return output
+
+        return execute_recursive(dag)
+
+    def get_stats(self) -> DatasetStats:
+        return self._stats
+
+
+def _naive_run_until_complete(op: PhysicalOperator) -> List[RefBundle]:
+    """Run this operator until completion, assuming all inputs have been submitted.
+
+    Args:
+        op: The operator to run.
+
+    Returns:
+        The list of output ref bundles for the operator.
+    """
+    output = []
+    tasks = op.get_work_refs()
+    if tasks:
+        bar = ProgressBar(op.name, total=op.num_outputs_total())
+        while tasks:
+            done, _ = ray.wait(
+                tasks, num_returns=len(tasks), fetch_local=True, timeout=0.1
+            )
+            for ready in done:
+                op.notify_work_completed(ready)
+            tasks = op.get_work_refs()
+            while op.has_next():
+                bar.update(1)
+                output.append(op.get_next())
+        bar.close()
+    # An operator is finished only after it has no remaining work as well as no
+    # remaining outputs.
+    while op.has_next():
+        output.append(op.get_next())
```

Comment:
I can see where you're coming from, but I believe it is a little premature to add this assumption. I'd prefer to keep this code as is until we see there is a concrete problem, since it is more defensively written with less assumptions on operator internals.

Reply:
>since it is more defensively written with less assumptions on operator internals.

I totally see the defensive approach here of always draining outputs even after the work task loop, and we can leave this as is since this isn't that important, but I want to call out that where we disagree is that I think this is about an important API contract between the executor and operators, not about operator internals.

IMO, this is about the executor + operator contract for when the executor can expect the operator to (1) have new outputs, and (2) be done producing outputs. Without a sensible contract (e.g. based on events that the executor can observe, like work tasks finishing), the executor won't know when it should invoke another `while op.has_next()` consuming loop, and won't know when the operator is done producing.

For example, does the executor support an operator that produces new output every 10 seconds without any internal work tasks? Definitely not!

My contention is that a sensible contract is:
1. If an operator has internal work tasks, then the executor can assume that outputs are produced by those work tasks finishing, and the operator therefore only produces more outputs after a task completes (i.e. after op.notify_work_completed() is called).
2. If an operator has no internal work tasks, then it has a static set of outputs that can be immediately consumed in a single pass.

And what I was pushing for above is explicitly representing this two-case contract in code. But yeah, definitely a pretty nitpicky change, not a high priority, and I think that the contract is clear enough as it is!